runtime-modules/content-directory/src/lib.rs

```diff
@@ -0,0 +1,1645 @@
+            EntityPermissions::<T>::ensure_group_can_remove_entity(access_level)?;
+
+            // Ensure there is no property values pointing to the given entity
+            Self::ensure_rc_is_zero(entity_id)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            Self::complete_entity_removal(entity_id);
+
+            Ok(())
+        }
+
+        pub fn add_schema_support_to_entity(
+            origin,
+            actor: Actor<T>,
+            entity_id: T::EntityId,
+            schema_id: SchemaId,
+            property_values: BTreeMap<PropertyId, PropertyValue<T>>
+        ) -> dispatch::Result {
+            let account_id = ensure_signed(origin)?;
+
+            Self::ensure_known_entity_id(entity_id)?;
+
+            let (entity, _) = Self::get_entity_and_access_level(account_id, entity_id, actor)?;
+
+            let class = Self::class_by_id(entity.class_id);
+
+            // Check that schema_id is a valid index of class schemas vector:
+            class.ensure_schema_id_exists(schema_id)?;
+
+            // Ensure class schema is active
+            class.ensure_schema_is_active(schema_id)?;
+
+            // Check that schema id is not yet added to this entity
+            entity.ensure_schema_id_is_not_added(schema_id)?;
+
+            let class_schema_opt = class.schemas.get(schema_id as usize);
+            let schema_prop_ids = class_schema_opt.unwrap().get_properties();
+
+            let current_entity_values = entity.values.clone();
+            let mut appended_entity_values = entity.values.clone();
+            let mut entities_rc_to_increment_vec = vec![];
+
+            for prop_id in schema_prop_ids.iter() {
+                if current_entity_values.contains_key(prop_id) {
+                    // A property is already added to the entity and cannot be updated
+                    // while adding a schema support to this entity.
+                    continue;
+                }
+                Self::add_new_property_value(&class, &entity, *prop_id, &property_values, &mut entities_rc_to_increment_vec, &mut appended_entity_values)?;
+            }
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            <EntityById<T>>::mutate(entity_id, |entity| {
+                // Add a new schema to the list of schemas supported by this entity.
+                entity.supported_schemas.insert(schema_id);
+
+                // Update entity values only if new properties have been added.
+                if appended_entity_values.len() > entity.values.len() {
+                    entity.values = appended_entity_values;
+                }
+            });
+            entities_rc_to_increment_vec
+                .iter()
+                .for_each(|entities_rc_to_increment| {
+                    Self::increment_entities_rc(entities_rc_to_increment);
+                });
+
+            Ok(())
+        }
+
+        pub fn update_entity_property_values(
```

Comment:
Same applies here, this is too complex too, even _if it works_, you have to find a way to simplify.

For example, you are mixing very late checks that actor is not locked out of properties at the same stage as computing referencing counting side effect vectors at L1396. It took me a long time to see that this check was in fact present, and needed. It would be better if you try to compute in simple steps, with local processing that does one simple thing at a time.

Reply:
Simplified here a lot https://github.com/iorveth/substrate-runtime-joystream/commit/fb5ab12c0d5e0ae127c0ccdfb6c837a96ab5d9a6